// @a24z/core-library/dist/config/loader.js
import { join, resolve } from 'path';
import { CONFIG_FILENAMES, DEFAULT_CONFIG } from './schema.js';
export class ConfigLoader {
    configCache = new Map();
    fsAdapter;
    constructor(fsAdapter) {
        this.fsAdapter = fsAdapter;
    }
    findConfigFile(startDir = process.cwd()) {
        let currentDir = resolve(startDir);
        const root = resolve('/');
        while (currentDir !== root) {
            for (const filename of CONFIG_FILENAMES) {
                const configPath = join(currentDir, filename);
                if (this.fsAdapter.exists(configPath)) {
                    return configPath;
                }
            }
            const parentDir = resolve(currentDir, '..');
            if (parentDir === currentDir)
                break;
            currentDir = parentDir;
        }
        return null;
    }
    loadConfig(configPath) {
        const path = configPath || this.findConfigFile();
        if (!path) {
            return null;
        }
        if (this.configCache.has(path)) {
            return this.configCache.get(path);
        }
        try {
            const content = this.fsAdapter.readFile(path);
            const config = JSON.parse(content);
            const merged = this.mergeWithDefaults(config);
            this.configCache.set(path, merged);
            return merged;
        }
        catch (error) {
            console.error(`Failed to load config from ${path}:`, error);
            return null;
        }
    }
    mergeWithDefaults(config) {
        return {
            ...DEFAULT_CONFIG,
            ...config,
            context: {
                ...DEFAULT_CONFIG.context,
                ...config.context,
                patterns: {
                    ...DEFAULT_CONFIG.context?.patterns,
                    ...config.context?.patterns,
                },
            },
            reporting: {
                ...DEFAULT_CONFIG.reporting,
                ...config.reporting,
            },
        };
    }
    clearCache() {
        this.configCache.clear();
    }
}
