// @a24z/core-library/dist/config/validator.js
export class ConfigValidator {
    validate(config) {
        const errors = [];
        const warnings = [];
        if (!this.isObject(config)) {
            errors.push({
                path: 'root',
                message: 'Configuration must be a valid JSON object',
            });
            return { valid: false, errors, warnings };
        }
        const cfg = config;
        // Validate version
        if (!cfg.version) {
            warnings.push({
                path: 'version',
                message: 'Version field is missing (defaulting to 1.0.0)',
                suggestion: 'Add "version": "1.0.0" to your config',
            });
        }
        else if (cfg.version !== '1.0.0') {
            errors.push({
                path: 'version',
                message: `Unsupported version: ${cfg.version}. Expected "1.0.0"`,
                value: cfg.version,
            });
        }
        // Validate project (optional)
        if (cfg.project !== undefined) {
            if (!this.isObject(cfg.project)) {
                errors.push({
                    path: 'project',
                    message: 'Project must be an object',
                    value: cfg.project,
                });
            }
            else {
                const project = cfg.project;
                if (project.name && typeof project.name !== 'string') {
                    errors.push({
                        path: 'project.name',
                        message: 'Project name must be a string',
                        value: project.name,
                    });
                }
                if (project.type !== undefined) {
                    const validTypes = ['library', 'application', 'monorepo', 'service'];
                    if (!validTypes.includes(project.type)) {
                        errors.push({
                            path: 'project.type',
                            message: `Invalid project type. Must be one of: ${validTypes.join(', ')}`,
                            value: project.type,
                        });
                    }
                }
            }
        }
        // Validate context
        if (cfg.context !== undefined) {
            if (!this.isObject(cfg.context)) {
                errors.push({
                    path: 'context',
                    message: 'Context must be an object',
                    value: cfg.context,
                });
            }
            else {
                this.validateContext(cfg.context, errors);
            }
        }
        // Validate reporting
        if (cfg.reporting !== undefined) {
            if (!this.isObject(cfg.reporting)) {
                errors.push({
                    path: 'reporting',
                    message: 'Reporting configuration must be an object',
                    value: cfg.reporting,
                });
            }
            else {
                this.validateReporting(cfg.reporting, errors);
            }
        }
        // Check for unknown top-level keys
        const knownKeys = ['$schema', 'version', 'project', 'context', 'reporting'];
        const unknownKeys = Object.keys(cfg).filter(key => !knownKeys.includes(key));
        if (unknownKeys.length > 0) {
            warnings.push({
                path: 'root',
                message: `Unknown configuration keys: ${unknownKeys.join(', ')}`,
                suggestion: 'Remove unknown keys or check for typos',
            });
        }
        return {
            valid: errors.length === 0,
            errors,
            warnings,
        };
    }
    validateContext(context, errors) {
        // Validate rules
        if (context.rules !== undefined) {
            if (!Array.isArray(context.rules)) {
                errors.push({
                    path: 'context.rules',
                    message: 'Rules must be an array',
                    value: context.rules,
                });
            }
            else {
                context.rules.forEach((rule, index) => {
                    if (!this.isObject(rule)) {
                        errors.push({
                            path: `context.rules[${index}]`,
                            message: 'Each rule must be an object',
                        });
                    }
                    else {
                        const r = rule;
                        if (!r.id) {
                            errors.push({
                                path: `context.rules[${index}].id`,
                                message: 'Rule id is required',
                            });
                        }
                        if (!r.name) {
                            errors.push({
                                path: `context.rules[${index}].name`,
                                message: 'Rule name is required',
                            });
                        }
                        if (!r.severity) {
                            errors.push({
                                path: `context.rules[${index}].severity`,
                                message: 'Rule severity is required',
                            });
                        }
                        else {
                            const validSeverities = ['error', 'warning', 'info'];
                            if (!validSeverities.includes(r.severity)) {
                                errors.push({
                                    path: `context.rules[${index}].severity`,
                                    message: `Invalid severity. Must be one of: ${validSeverities.join(', ')}`,
                                    value: r.severity,
                                });
                            }
                        }
                        if (r.enabled !== undefined && typeof r.enabled !== 'boolean') {
                            errors.push({
                                path: `context.rules[${index}].enabled`,
                                message: 'Rule enabled must be a boolean',
                                value: r.enabled,
                            });
                        }
                    }
                });
            }
        }
        // Validate patterns
        if (context.patterns !== undefined) {
            if (!this.isObject(context.patterns)) {
                errors.push({
                    path: 'context.patterns',
                    message: 'Patterns must be an object',
                    value: context.patterns,
                });
            }
            else {
                const patterns = context.patterns;
                if (patterns.include !== undefined && !Array.isArray(patterns.include)) {
                    errors.push({
                        path: 'context.patterns.include',
                        message: 'Include patterns must be an array of strings',
                        value: patterns.include,
                    });
                }
                if (patterns.exclude !== undefined && !Array.isArray(patterns.exclude)) {
                    errors.push({
                        path: 'context.patterns.exclude',
                        message: 'Exclude patterns must be an array of strings',
                        value: patterns.exclude,
                    });
                }
            }
        }
        // Validate boolean fields
        if (context.useGitignore !== undefined && typeof context.useGitignore !== 'boolean') {
            errors.push({
                path: 'context.useGitignore',
                message: 'useGitignore must be a boolean',
                value: context.useGitignore,
            });
        }
        if (context.followSymlinks !== undefined && typeof context.followSymlinks !== 'boolean') {
            errors.push({
                path: 'context.followSymlinks',
                message: 'followSymlinks must be a boolean',
                value: context.followSymlinks,
            });
        }
        if (context.maxDepth !== undefined && typeof context.maxDepth !== 'number') {
            errors.push({
                path: 'context.maxDepth',
                message: 'maxDepth must be a number',
                value: context.maxDepth,
            });
        }
    }
    validateReporting(reporting, errors) {
        if (reporting.output !== undefined) {
            const validOutputs = ['console', 'file', 'both'];
            if (!validOutputs.includes(reporting.output)) {
                errors.push({
                    path: 'reporting.output',
                    message: `Invalid output. Must be one of: ${validOutputs.join(', ')}`,
                    value: reporting.output,
                });
            }
        }
        if (reporting.format !== undefined) {
            const validFormats = ['text', 'json', 'html'];
            if (!validFormats.includes(reporting.format)) {
                errors.push({
                    path: 'reporting.format',
                    message: `Invalid format. Must be one of: ${validFormats.join(', ')}`,
                    value: reporting.format,
                });
            }
        }
        if (reporting.verbose !== undefined && typeof reporting.verbose !== 'boolean') {
            errors.push({
                path: 'reporting.verbose',
                message: 'verbose must be a boolean',
                value: reporting.verbose,
            });
        }
    }
    isObject(value) {
        return typeof value === 'object' && value !== null && !Array.isArray(value);
    }
}
