// @a24z/core-library/dist/pure-core/types/summary.js
/**
 * Summary types for CodebaseViews
 * Lightweight representations for listing and preview purposes
 */
/**
 * Extract a summary from a full CodebaseView
 * Calculates grid dimensions and cell count
 */
export function extractCodebaseViewSummary(view) {
    let maxRow = 0;
    let maxCol = 0;
    const cellCount = Object.keys(view.cells).length;
    // Calculate grid dimensions from cell coordinates
    for (const cell of Object.values(view.cells)) {
        if (cell.coordinates && Array.isArray(cell.coordinates)) {
            const [row, col] = cell.coordinates;
            maxRow = Math.max(maxRow, row);
            maxCol = Math.max(maxCol, col);
        }
    }
    return {
        id: view.id,
        name: view.name,
        description: view.description || '',
        cellCount,
        gridSize: [maxRow + 1, maxCol + 1], // +1 because coordinates are 0-indexed
        overviewPath: view.overviewPath || '',
        category: view.category || 'other', // Default to 'other' for backward compatibility
        displayOrder: view.displayOrder || 0, // Default to 0 for backward compatibility
    };
}
/**
 * Create summaries for multiple views
 */
export function extractCodebaseViewSummaries(views) {
    return views.map(extractCodebaseViewSummary);
}
