// @a24z/core-library/dist/pure-core/types/validation.js
/**
 * Validation types for a24z-memory
 * These types are used for note validation and custom error messages
 */
// ============================================================================
// Default Messages
// ============================================================================
/**
 * Default validation message templates
 */
export const DEFAULT_VALIDATION_MESSAGES = {
    noteTooLong: (data) => `Note content is too long (${data.actual.toLocaleString()} characters, ${data.percentage}% of limit). ` +
        `Maximum allowed: ${data.limit.toLocaleString()} characters. ` +
        `You are ${data.overBy.toLocaleString()} characters over the limit. ` +
        `💡 Tip: Consider splitting this into multiple focused notes.`,
    tooManyTags: (data) => `Note has too many tags (${data.actual}). Maximum allowed: ${data.limit}`,
    tooManyAnchors: (data) => `Note has too many anchors (${data.actual}). Maximum allowed: ${data.limit}`,
    invalidTags: (data) => `The following tags are not allowed: ${data.invalidTags.join(', ')}. ` +
        `Allowed tags: ${data.allowedTags.join(', ')}`,
    invalidType: (data) => `Invalid note type: ${data.type}. Allowed types: ${data.allowedTypes.join(', ')}`,
    anchorOutsideRepo: (data) => `Anchor is outside the repository: ${data.anchor}`,
    missingAnchors: (data) => `Note must have at least one anchor (currently has ${data.actual})`,
};
// ============================================================================
// Validation Message Formatter
// ============================================================================
/**
 * Formatter class for validation messages
 * Allows customization of validation error messages
 */
export class ValidationMessageFormatter {
    messages;
    constructor(overrides) {
        this.messages = { ...DEFAULT_VALIDATION_MESSAGES };
        if (overrides) {
            Object.assign(this.messages, overrides);
        }
    }
    /**
     * Format a validation error with the appropriate message
     */
    format(type, data) {
        const formatter = this.messages[type];
        if (!formatter) {
            return `Validation error: ${type}`;
        }
        return formatter(data);
    }
    /**
     * Get all available message types
     */
    getAvailableTypes() {
        return Object.keys(this.messages);
    }
}
