// @principal-ai/repository-abstraction/src/github-tools.ts
// Repository information for resolving relative URLs
export interface GitHubRepositoryInfo {
  /** GitHub repository owner/organization */
  owner: string;
  /** Repository name */
  repo: string;
  /** Branch/ref to use for raw URLs (defaults to 'main' if not specified) */
  branch?: string;
  /** Base path within the repository (for files in subdirectories) */
  basePath?: string;
}
/**
 * Parses a GitHub URL and extracts repository information
 */
export function parseGitHubUrl(url: string): GitHubRepositoryInfo | null {
  try {
    // Handle various GitHub URL formats:
    // https://github.com/owner/repo
    // https://github.com/owner/repo/blob/branch/path/to/file.md
    // https://github.com/owner/repo/tree/branch/path/to/directory
    // https://raw.githubusercontent.com/owner/repo/branch/path/to/file.md

    const githubUrl = url.trim();

    // Handle raw.githubusercontent.com URLs
    const rawMatch = githubUrl.match(
      /^https?:\/\/raw\.githubusercontent\.com\/([^/]+)\/([^/]+)\/([^/]+)\/(.*)$/,
    );
    if (rawMatch) {
      const [, owner, repo, branch, filePath] = rawMatch;
      return {
        owner,
        repo,
        branch,
        basePath: filePath.includes('/')
          ? filePath.substring(0, filePath.lastIndexOf('/'))
          : undefined,
      };
    }

    // Handle github.com URLs
    const githubMatch = githubUrl.match(
      /^https?:\/\/github\.com\/([^/]+)\/([^/]+)(?:\/(blob|tree)\/([^/]+)\/(.*))?$/,
    );
    if (githubMatch) {
      const [, owner, repo, , branch = 'main', filePath = ''] = githubMatch;
      return {
        owner,
        repo,
        branch,
        basePath:
          filePath && filePath.includes('/')
            ? filePath.substring(0, filePath.lastIndexOf('/'))
            : undefined,
      };
    }

    // Handle owner/repo format
    const shortMatch = githubUrl.match(/^([^/]+)\/([^/]+)$/);
    if (shortMatch) {
      const [, owner, repo] = shortMatch;
      return {
        owner,
        repo,
        branch: 'main',
        basePath: undefined,
      };
    }

    return null;
  } catch {
    return null;
  }
}
