// @a24z/core-library/dist/projects-core/AlexandriaOutpostManager.js
import { ProjectRegistryStore } from './ProjectRegistryStore.js';
import { MemoryPalace } from '../MemoryPalace.js';
import { extractCodebaseViewSummary } from '../pure-core/types/summary.js';
import { homedir } from 'os';
export class AlexandriaOutpostManager {
    fsAdapter;
    projectRegistry;
    constructor(fsAdapter) {
        this.fsAdapter = fsAdapter;
        // Create the ProjectRegistryStore internally with the user's home directory
        this.projectRegistry = new ProjectRegistryStore(fsAdapter, homedir());
    }
    async getAllRepositories() {
        // Get all registered projects from existing registry
        const entries = this.projectRegistry.listProjects();
        // Transform each to API format
        const repositories = await Promise.all(entries.map(entry => this.transformToRepository(entry)));
        return repositories.filter(repo => repo !== null);
    }
    async getRepository(name) {
        const entry = this.projectRegistry.getProject(name);
        if (!entry)
            return null;
        return this.transformToRepository(entry);
    }
    async registerRepository(name, path) {
        // Use existing registry's register method
        this.projectRegistry.registerProject(name, path);
        // Return the transformed repository
        const entry = this.projectRegistry.getProject(name);
        if (!entry) {
            throw new Error(`Failed to register repository ${name}`);
        }
        return this.transformToRepository(entry);
    }
    async getRepositoryByPath(path) {
        // Find repository by path
        const entries = this.projectRegistry.listProjects();
        const entry = entries.find(e => e.path === path);
        if (!entry)
            return null;
        return this.transformToRepository(entry);
    }
    getRepositoryCount() {
        return this.projectRegistry.listProjects().length;
    }
    async transformToRepository(entry) {
        // Load views if not already loaded
        let views = entry.views || [];
        if (views.length === 0) {
            try {
                // Create a new MemoryPalace instance for this repository
                const memoryPalace = new MemoryPalace(entry.path, this.fsAdapter);
                // Get views from the memory palace
                views = memoryPalace.listViews().map(v => extractCodebaseViewSummary(v));
            }
            catch (error) {
                // If we can't load views, continue with empty array
                console.debug(`Could not load views for ${entry.name}:`, error);
                views = [];
            }
        }
        // Extract owner from remote URL if available
        const owner = this.extractOwner(entry.remoteUrl);
        // Build the repository data according to AlexandriaRepository type
        const repo = {
            name: entry.name,
            remoteUrl: entry.remoteUrl,
            registeredAt: entry.registeredAt,
            hasViews: views.length > 0,
            viewCount: views.length,
            views,
            // Only include github if we have github data or can construct it
            github: entry.github || (owner ? {
                id: `${owner}/${entry.name}`,
                owner: owner,
                name: entry.name,
                stars: 0,
                lastUpdated: new Date().toISOString()
            } : undefined)
        };
        return repo;
    }
    extractOwner(remoteUrl) {
        if (!remoteUrl)
            return null;
        // Extract owner from git URL
        const match = remoteUrl.match(/github\.com[:/]([^/]+)/);
        return match ? match[1] : null;
    }
}
