// @a24z/core-library/dist/projects-core/ProjectRegistryStore.js
/**
 * Pure ProjectRegistryStore - Platform-agnostic project registry management
 *
 * Manages a registry of local project paths with their git remotes
 */
export class ProjectRegistryStore {
    fs;
    registryPath;
    constructor(fileSystemAdapter, homeDir) {
        this.fs = fileSystemAdapter;
        this.registryPath = this.fs.join(homeDir, '.alexandria', 'projects.json');
    }
    /**
     * Ensure the registry directory exists
     */
    ensureRegistryDir() {
        const registryDir = this.fs.dirname(this.registryPath);
        if (!this.fs.exists(registryDir)) {
            this.fs.createDir(registryDir);
        }
    }
    /**
     * Load the project registry
     */
    loadRegistry() {
        this.ensureRegistryDir();
        if (!this.fs.exists(this.registryPath)) {
            return {
                version: '1.0.0',
                projects: [],
            };
        }
        try {
            const content = this.fs.readFile(this.registryPath);
            return JSON.parse(content);
        }
        catch {
            return {
                version: '1.0.0',
                projects: [],
            };
        }
    }
    /**
     * Save the project registry
     */
    saveRegistry(registry) {
        this.ensureRegistryDir();
        this.fs.writeFile(this.registryPath, JSON.stringify(registry, null, 2));
    }
    /**
     * Register a new project
     */
    registerProject(name, projectPath, remoteUrl) {
        const registry = this.loadRegistry();
        // Check if name already exists
        if (registry.projects.some((p) => p.name === name)) {
            throw new Error(`Project with name '${name}' already exists`);
        }
        // Check if path already registered
        const existingProject = registry.projects.find((p) => p.path === projectPath);
        if (existingProject) {
            throw new Error(`Path already registered as '${existingProject.name}'`);
        }
        const entry = {
            name,
            path: projectPath,
            remoteUrl,
            registeredAt: new Date().toISOString(),
            github: undefined, // Will be populated when fetching from GitHub
            hasViews: false, // Will be updated when scanning for views
            viewCount: 0,
            views: [],
        };
        registry.projects.push(entry);
        this.saveRegistry(registry);
    }
    /**
     * List all registered projects
     */
    listProjects() {
        const registry = this.loadRegistry();
        return registry.projects;
    }
    /**
     * Get a project by name
     */
    getProject(name) {
        const registry = this.loadRegistry();
        return registry.projects.find((p) => p.name === name);
    }
    /**
     * Remove a project from the registry
     */
    removeProject(name) {
        const registry = this.loadRegistry();
        const initialLength = registry.projects.length;
        registry.projects = registry.projects.filter((p) => p.name !== name);
        if (registry.projects.length < initialLength) {
            this.saveRegistry(registry);
            return true;
        }
        return false;
    }
    /**
     * Update a project's path or remote URL
     */
    updateProject(name, updates) {
        const registry = this.loadRegistry();
        const project = registry.projects.find((p) => p.name === name);
        if (!project) {
            throw new Error(`Project '${name}' not found`);
        }
        if (updates.path) {
            // Check if new path is already registered
            const existingWithPath = registry.projects.find((p) => p.path === updates.path && p.name !== name);
            if (existingWithPath) {
                throw new Error(`Path already registered as '${existingWithPath.name}'`);
            }
            project.path = updates.path;
        }
        if ('remoteUrl' in updates) {
            if (updates.remoteUrl === undefined) {
                delete project.remoteUrl;
            }
            else {
                project.remoteUrl = updates.remoteUrl;
            }
        }
        this.saveRegistry(registry);
    }
}
