// @principal-ai/repository-abstraction/dist/fileTreeAdapters.js
"use strict";
/**
 * FileTree adapter utilities
 * Convert various file representations into FileTree format
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFileTreeFromPaths = createFileTreeFromPaths;
exports.mergeFileTrees = mergeFileTrees;
/**
 * Convert a simple file list to FileTree format
 * Useful when you only have file paths without metadata
 *
 * @param files - Array of file paths (e.g., ['src/index.ts', 'package.json'])
 * @param rootPath - Root directory path for calculating relative paths
 * @returns A minimal but valid FileTree structure
 */
function createFileTreeFromPaths(files, rootPath = '') {
    // Normalize root path
    const normalizedRoot = rootPath.endsWith('/') ? rootPath.slice(0, -1) : rootPath;
    // Build tree structure from file paths
    const tree = new Map();
    const allFiles = [];
    const allDirectories = [];
    // First, create all directories
    const directoryPaths = new Set();
    files.forEach(filePath => {
        const parts = filePath.split('/');
        // Build all parent directory paths
        for (let i = 1; i < parts.length; i++) {
            directoryPaths.add(parts.slice(0, i).join('/'));
        }
    });
    // Create DirectoryInfo objects and add to tree
    Array.from(directoryPaths).forEach(dirPath => {
        const parts = dirPath.split('/');
        const name = parts[parts.length - 1] || '';
        const depth = parts.length - 1;
        const dirInfo = {
            path: dirPath,
            name,
            children: [],
            fileCount: 0,
            totalSize: 0,
            depth,
            relativePath: calculateRelativePath(dirPath, normalizedRoot),
        };
        tree.set(dirPath, dirInfo);
        allDirectories.push(dirInfo);
    });
    // Then, create all files and add them to their parent directories
    files.forEach(filePath => {
        const parts = filePath.split('/');
        const name = parts[parts.length - 1] || '';
        const extension = extractExtension(name);
        const parentPath = parts.slice(0, -1).join('/');
        const fileInfo = {
            path: filePath,
            name,
            extension,
            size: 0,
            lastModified: new Date(),
            isDirectory: false,
            relativePath: calculateRelativePath(filePath, normalizedRoot),
        };
        allFiles.push(fileInfo);
        // Add file to its parent directory
        const parent = tree.get(parentPath);
        if (parent && 'children' in parent) {
            parent.children.push(fileInfo);
            parent.fileCount++;
        }
    });
    // Build the tree structure by linking directories to their parents
    Array.from(directoryPaths)
        .sort((a, b) => b.length - a.length)
        .forEach(dirPath => {
        const parts = dirPath.split('/');
        if (parts.length > 1) {
            const parentPath = parts.slice(0, -1).join('/');
            const parent = tree.get(parentPath);
            const current = tree.get(dirPath);
            if (parent && current && 'children' in parent) {
                parent.children.push(current);
            }
        }
    });
    // Calculate file type distribution
    const buildingTypeDistribution = {};
    const directoryTypeDistribution = {};
    const combinedTypeDistribution = {};
    allFiles.forEach(file => {
        const ext = file.extension || 'no-extension';
        buildingTypeDistribution[ext] = (buildingTypeDistribution[ext] || 0) + 1;
        combinedTypeDistribution[ext] = (combinedTypeDistribution[ext] || 0) + 1;
    });
    allDirectories.forEach(dir => {
        const type = classifyDirectory(dir.name);
        directoryTypeDistribution[type] = (directoryTypeDistribution[type] || 0) + 1;
        combinedTypeDistribution[`dir:${type}`] = (combinedTypeDistribution[`dir:${type}`] || 0) + 1;
    });
    // Find or create root directory
    const rootName = normalizedRoot ? normalizedRoot.split('/').pop() || 'root' : 'root';
    const maxDepth = allDirectories.length > 0 ? Math.max(...allDirectories.map(d => d.depth)) : 0;
    // Find top-level directories and files (those with only one path segment)
    const topLevelChildren = [];
    // Add top-level directories
    allDirectories.forEach(dir => {
        const parts = dir.path.split('/');
        if (parts.length === 1) {
            topLevelChildren.push(dir);
        }
    });
    // Add top-level files (if any)
    allFiles.forEach(file => {
        const parts = file.path.split('/');
        if (parts.length === 1) {
            topLevelChildren.push(file);
        }
    });
    const root = {
        path: normalizedRoot || '.',
        name: rootName,
        children: topLevelChildren,
        fileCount: allFiles.length,
        totalSize: 0,
        depth: 0,
        relativePath: '',
    };
    return {
        sha: generateSha(files),
        root,
        allFiles,
        allDirectories,
        stats: {
            totalFiles: allFiles.length,
            totalDirectories: allDirectories.length,
            totalSize: 0, // Would need file system access for real size
            maxDepth,
            buildingTypeDistribution,
            directoryTypeDistribution,
            combinedTypeDistribution,
        },
    };
}
/**
 * Merge multiple FileTrees into one
 * Useful for combining trees from different sources
 */
function mergeFileTrees(...trees) {
    if (trees.length === 0) {
        return createFileTreeFromPaths([], '');
    }
    if (trees.length === 1) {
        return trees[0];
    }
    // Combine all files and directories
    const allFiles = [];
    const allDirectories = [];
    const filePathSet = new Set();
    const dirPathSet = new Set();
    trees.forEach(tree => {
        tree.allFiles.forEach(file => {
            if (!filePathSet.has(file.path)) {
                filePathSet.add(file.path);
                allFiles.push(file);
            }
        });
        tree.allDirectories.forEach(dir => {
            if (!dirPathSet.has(dir.path)) {
                dirPathSet.add(dir.path);
                allDirectories.push(dir);
            }
        });
    });
    // Recalculate stats
    const buildingTypeDistribution = {};
    const directoryTypeDistribution = {};
    const combinedTypeDistribution = {};
    allFiles.forEach(file => {
        const ext = file.extension || 'no-extension';
        buildingTypeDistribution[ext] = (buildingTypeDistribution[ext] || 0) + 1;
        combinedTypeDistribution[ext] = (combinedTypeDistribution[ext] || 0) + 1;
    });
    allDirectories.forEach(dir => {
        const type = classifyDirectory(dir.name);
        directoryTypeDistribution[type] = (directoryTypeDistribution[type] || 0) + 1;
        combinedTypeDistribution[`dir:${type}`] = (combinedTypeDistribution[`dir:${type}`] || 0) + 1;
    });
    const totalSize = allFiles.reduce((sum, f) => sum + f.size, 0);
    const maxDepth = allDirectories.length > 0 ? Math.max(...allDirectories.map(d => d.depth)) : 0;
    // Use first tree's root as base
    const root = trees[0].root;
    root.fileCount = allFiles.length;
    root.totalSize = totalSize;
    return {
        sha: `merged-${Date.now()}`,
        root,
        allFiles,
        allDirectories,
        stats: {
            totalFiles: allFiles.length,
            totalDirectories: allDirectories.length,
            totalSize,
            maxDepth,
            buildingTypeDistribution,
            directoryTypeDistribution,
            combinedTypeDistribution,
        },
    };
}
// Helper functions
function extractExtension(filename) {
    if (!filename.includes('.'))
        return '';
    const parts = filename.split('.');
    return '.' + parts[parts.length - 1].toLowerCase();
}
function calculateRelativePath(fullPath, rootPath) {
    // If no root path or root path is empty/current directory, return the full path as relative
    if (!rootPath || rootPath === '' || rootPath === '.') {
        return fullPath;
    }
    if (fullPath === rootPath)
        return '';
    if (fullPath.startsWith(rootPath)) {
        const relative = fullPath.substring(rootPath.length);
        return relative.startsWith('/') ? relative.substring(1) : relative;
    }
    return fullPath;
}
function classifyDirectory(name) {
    const lowerName = name.toLowerCase();
    // Common directory classifications
    if (lowerName === 'src' || lowerName === 'source')
        return 'source';
    if (lowerName === 'test' || lowerName === 'tests' || lowerName === '__tests__')
        return 'test';
    if (lowerName === 'docs' || lowerName === 'documentation')
        return 'docs';
    if (lowerName === 'lib' || lowerName === 'libs')
        return 'library';
    if (lowerName === 'dist' || lowerName === 'build' || lowerName === 'out')
        return 'build';
    if (lowerName === 'node_modules' || lowerName === 'vendor')
        return 'dependencies';
    if (lowerName === 'public' || lowerName === 'static' || lowerName === 'assets')
        return 'assets';
    if (lowerName === 'config' || lowerName === 'configs' || lowerName === '.config')
        return 'config';
    if (lowerName.startsWith('.'))
        return 'hidden';
    return 'other';
}
function generateSha(files) {
    // Simple hash based on file count and timestamp
    // In production, you might want to use a real hash of file contents
    const fileStr = files.sort().join(',');
    const simpleHash = fileStr.split('').reduce((acc, char) => {
        return (acc << 5) - acc + char.charCodeAt(0);
    }, 0);
    return `tree-${Math.abs(simpleHash)}-${Date.now()}`;
}
