// @a24z/core-library/dist/pure-core/stores/AnchoredNotesStore.js
/**
 * Pure AnchoredNotesStore - Platform-agnostic note storage
 *
 * This version uses dependency injection with FileSystemAdapter to work in any environment
 * No Node.js dependencies - can run in browsers, Deno, Bun, or anywhere JavaScript runs
 */
import { A24zConfigurationStore } from './A24zConfigurationStore.js';
import { ALEXANDRIA_DIRS } from '../../constants/paths.js';
// ============================================================================
// Pure AnchoredNotesStore Class
// ============================================================================
export class AnchoredNotesStore {
    fs;
    configStore;
    alexandriaPath;
    notesDir;
    constructor(fileSystemAdapter, alexandriaPath) {
        this.fs = fileSystemAdapter;
        this.alexandriaPath = alexandriaPath;
        this.notesDir = this.fs.join(alexandriaPath, 'notes');
        // Pass the same alexandriaPath to config store
        this.configStore = new A24zConfigurationStore(fileSystemAdapter, alexandriaPath);
        // Ensure notes directory exists
        this.fs.createDir(this.notesDir);
    }
    // ============================================================================
    // Path Utilities (pure functions that work with any path format)
    // ============================================================================
    /**
     * Get the notes directory path
     */
    getNotesDir() {
        // Use the pre-computed notes directory
        return this.notesDir;
    }
    /**
     * Get the tags directory path
     */
    getTagsDir() {
        return this.fs.join(this.alexandriaPath, 'tags');
    }
    /**
     * Get path for a specific note file using date-based directory structure
     */
    getNotePath(repositoryRootPath, noteId, timestamp) {
        const date = new Date(timestamp);
        const year = date.getFullYear();
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const noteDir = this.fs.join(this.getNotesDir(), year.toString(), month);
        return this.fs.join(noteDir, `${noteId}.json`);
    }
    /**
     * Ensure the date-based notes directory exists
     */
    ensureNotesDir(repositoryRootPath, timestamp) {
        const date = new Date(timestamp);
        const year = date.getFullYear();
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const noteDir = this.fs.join(this.getNotesDir(), year.toString(), month);
        this.fs.createDir(noteDir);
    }
    /**
     * Generate a unique note ID
     */
    generateNoteId() {
        const timestamp = Date.now();
        const random = Math.random().toString(36).substr(2, 9);
        return `note-${timestamp}-${random}`;
    }
    // ============================================================================
    // Configuration Management (delegated to A24zConfigurationStore)
    // ============================================================================
    /**
     * Get repository configuration
     */
    getConfiguration() {
        return this.configStore.getConfiguration();
    }
    /**
     * Update repository configuration
     */
    updateConfiguration(repositoryRootPath, updates) {
        return this.configStore.updateConfiguration(updates);
    }
    /**
     * Enable or disable allowed tags enforcement
     */
    setEnforceAllowedTags(repositoryRootPath, enforce) {
        return this.configStore.setEnforceAllowedTags(enforce);
    }
    // ============================================================================
    // Note CRUD Operations
    // ============================================================================
    /**
     * Save a note to storage
     */
    saveNote(input) {
        const { note: noteContent, anchors, tags, codebaseViewId, metadata, directoryPath } = input;
        // Validate input
        const validation = this.validateNote({ note: noteContent, anchors, tags, codebaseViewId, metadata }, directoryPath);
        if (validation.length > 0) {
            throw new Error(`Validation failed: ${validation.map((v) => v.message).join(', ')}`);
        }
        // Create the note
        const noteId = this.generateNoteId();
        const timestamp = Date.now();
        const storedNote = {
            id: noteId,
            note: noteContent,
            anchors: [...anchors], // Clone to prevent mutations
            tags: [...tags], // Clone to prevent mutations
            metadata: { ...metadata }, // Clone to prevent mutations
            timestamp,
            codebaseViewId,
        };
        // Ensure date-based directory exists
        this.ensureNotesDir(directoryPath, timestamp);
        // Save to file using date-based path
        const notePath = this.getNotePath(directoryPath, noteId, timestamp);
        this.fs.writeFile(notePath, JSON.stringify(storedNote, null, 2));
        return {
            note: storedNote,
            path: this.fs.relative(directoryPath, notePath),
        };
    }
    /**
     * Get a note by its ID (searches through date-based directory structure)
     */
    getNoteById(repositoryRootPath, noteId) {
        const notesDir = this.getNotesDir();
        if (!this.fs.exists(notesDir)) {
            return null;
        }
        // Search through year/month directories
        const yearDirs = this.fs.readDir(notesDir).filter((item) => /^\d{4}$/.test(item));
        for (const year of yearDirs) {
            const yearDir = this.fs.join(notesDir, year);
            const monthDirs = this.fs.readDir(yearDir).filter((item) => /^\d{2}$/.test(item));
            for (const month of monthDirs) {
                const monthDir = this.fs.join(yearDir, month);
                const noteFilePath = this.fs.join(monthDir, `${noteId}.json`);
                if (this.fs.exists(noteFilePath)) {
                    try {
                        const content = this.fs.readFile(noteFilePath);
                        return JSON.parse(content);
                    }
                    catch (error) {
                        console.warn(`Failed to load note ${noteId}:`, error);
                        return null;
                    }
                }
            }
        }
        return null;
    }
    /**
     * Delete a note by its ID (searches through date-based directory structure)
     */
    deleteNoteById(repositoryRootPath, noteId) {
        const notesDir = this.getNotesDir();
        if (!this.fs.exists(notesDir)) {
            return false;
        }
        // Search through year/month directories
        const yearDirs = this.fs.readDir(notesDir).filter((item) => /^\d{4}$/.test(item));
        for (const year of yearDirs) {
            const yearDir = this.fs.join(notesDir, year);
            const monthDirs = this.fs.readDir(yearDir).filter((item) => /^\d{2}$/.test(item));
            for (const month of monthDirs) {
                const monthDir = this.fs.join(yearDir, month);
                const noteFilePath = this.fs.join(monthDir, `${noteId}.json`);
                if (this.fs.exists(noteFilePath)) {
                    try {
                        this.fs.deleteFile(noteFilePath);
                        return true;
                    }
                    catch (error) {
                        console.warn(`Failed to delete note ${noteId}:`, error);
                        return false;
                    }
                }
            }
        }
        return false;
    }
    /**
     * Get all notes for a specific path within a repository
     */
    getNotesForPath(repositoryRoot, relativePath, includeParentNotes = true) {
        // Get all notes from the repository
        const allNotes = this.readAllNotes(repositoryRoot);
        // Use the already-validated relative path for comparison
        const queryRelative = relativePath;
        // Filter and sort notes based on path matching
        const processedNotes = allNotes.map((noteWithPath) => {
            const note = noteWithPath.note;
            let isParent = false;
            // Check if any anchor matches the query path
            const matchesAnchor = note.anchors.some((anchor) => {
                // Normalize the anchor path
                const normalizedAnchor = anchor.replace(/\\/g, '/');
                const normalizedQuery = queryRelative.replace(/\\/g, '/');
                return (normalizedQuery === normalizedAnchor ||
                    normalizedQuery.startsWith(`${normalizedAnchor}/`) ||
                    normalizedAnchor.startsWith(`${normalizedQuery}/`));
            });
            // With ValidatedRelativePath, we know the query is always in the repository
            if (matchesAnchor) {
                isParent = false;
            }
            else {
                isParent = true; // If it doesn't match anchor, it must be a parent
            }
            // Calculate distance for sorting - use the relative path segments
            const distance = matchesAnchor
                ? 0
                : isParent
                    ? queryRelative.split('/').filter(Boolean).length
                    : 9999;
            return {
                note: { ...note, isParentDirectory: isParent, pathDistance: distance },
                path: noteWithPath.path,
            };
        });
        // Filter nulls and apply parent filter
        const filtered = processedNotes
            .filter((x) => x !== null)
            .filter((x) => (includeParentNotes ? true : !x.note.isParentDirectory));
        // Sort by distance and timestamp
        filtered.sort((a, b) => {
            return a.note.pathDistance - b.note.pathDistance || b.note.timestamp - a.note.timestamp;
        });
        // Return as AnchoredNoteWithPath array (remove the extra properties)
        return filtered.map((item) => ({
            note: item.note,
            path: item.path,
        }));
    }
    /**
     * Read all notes from a repository
     */
    readAllNotes(repositoryRootPath) {
        // Calculate the notes directory for the specific repository
        const repositoryAlexandriaPath = this.fs.join(repositoryRootPath, ALEXANDRIA_DIRS.PRIMARY);
        const notesDir = this.fs.join(repositoryAlexandriaPath, ALEXANDRIA_DIRS.NOTES);
        const notes = [];
        if (!this.fs.exists(notesDir)) {
            return [];
        }
        // Recursively read all .json files in the notes directory
        const readNotesRecursive = (dir) => {
            const entries = this.fs.readDir(dir);
            for (const entry of entries) {
                const fullPath = this.fs.join(dir, entry);
                // Check if it's a directory using the proper method
                const isDirectory = this.fs.isDirectory(fullPath);
                if (isDirectory) {
                    // Recurse into the directory
                    readNotesRecursive(fullPath);
                }
                else if (entry.endsWith('.json')) {
                    try {
                        const noteContent = this.fs.readFile(fullPath);
                        const note = JSON.parse(noteContent);
                        if (note && typeof note === 'object' && note.id) {
                            notes.push({
                                note,
                                path: this.fs.relative(repositoryRootPath, fullPath),
                            });
                        }
                    }
                    catch (error) {
                        // Skip invalid note files
                        console.warn(`Failed to read note file ${fullPath}:`, error);
                    }
                }
            }
        };
        readNotesRecursive(notesDir);
        return notes;
    }
    /**
     * Check for notes with stale anchors (files that no longer exist)
     */
    checkStaleAnchoredNotes(repositoryRootPath) {
        const notesWithPaths = this.readAllNotes(repositoryRootPath);
        const staleNotes = [];
        for (const noteWithPath of notesWithPaths) {
            const note = noteWithPath.note;
            const staleAnchors = [];
            const validAnchors = [];
            for (const anchor of note.anchors) {
                // Anchors are stored as relative paths to the repo root
                const anchorPath = this.fs.join(repositoryRootPath, anchor);
                if (this.fs.exists(anchorPath)) {
                    validAnchors.push(anchor);
                }
                else {
                    staleAnchors.push(anchor);
                }
            }
            // Only include notes that have at least one stale anchor
            if (staleAnchors.length > 0) {
                staleNotes.push({
                    note,
                    staleAnchors,
                    validAnchors,
                });
            }
        }
        return staleNotes;
    }
    // ============================================================================
    // Validation
    // ============================================================================
    /**
     * Validate a note against repository configuration
     * TODO: _repositoryRootPath to be used to validate anchors exist in the fs
     */
    validateNote(note, _repositoryRootPath) {
        const errors = [];
        const config = this.getConfiguration();
        // Check note length
        if (note.note.length > config.limits.noteMaxLength) {
            errors.push({
                type: 'noteTooLong',
                message: `Note is too long (${note.note.length} > ${config.limits.noteMaxLength})`,
                context: { actual: note.note.length, limit: config.limits.noteMaxLength },
            });
        }
        // Check tags count
        if (note.tags.length > config.limits.maxTagsPerNote) {
            errors.push({
                type: 'tooManyTags',
                message: `Too many tags (${note.tags.length} > ${config.limits.maxTagsPerNote})`,
                context: { actual: note.tags.length, limit: config.limits.maxTagsPerNote },
            });
        }
        // Check anchors count
        if (note.anchors.length > config.limits.maxAnchorsPerNote) {
            errors.push({
                type: 'tooManyAnchors',
                message: `Too many anchors (${note.anchors.length} > ${config.limits.maxAnchorsPerNote})`,
                context: { actual: note.anchors.length, limit: config.limits.maxAnchorsPerNote },
            });
        }
        // Check for missing anchors
        if (note.anchors.length === 0) {
            errors.push({
                type: 'missingAnchors',
                message: 'Notes must have at least one anchor',
                context: { actual: note.anchors.length },
            });
        }
        // Check for invalid anchor patterns (reject globs and wildcards)
        const invalidAnchors = [];
        const globPatterns = ['*', '?', '[', ']', '{', '}', '**'];
        for (const anchor of note.anchors) {
            // Check if anchor contains any glob pattern characters
            if (globPatterns.some((pattern) => anchor.includes(pattern))) {
                invalidAnchors.push(anchor);
            }
            // Also check for absolute paths (optional - you may want to keep this)
            if (anchor.startsWith('/')) {
                invalidAnchors.push(anchor);
            }
        }
        if (invalidAnchors.length > 0) {
            errors.push({
                type: 'invalidAnchor',
                message: `Invalid anchor paths detected. Anchors must be relative paths without glob patterns (*, ?, **, etc.). Invalid anchors: ${invalidAnchors.join(', ')}`,
                context: { invalidAnchors },
            });
        }
        // Check for invalid tags if enforcement is enabled
        if (config.tags?.enforceAllowedTags) {
            const allowedTags = this.getAllowedTags();
            if (allowedTags.enforced && allowedTags.tags.length > 0) {
                const invalidTags = note.tags.filter((tag) => !allowedTags.tags.includes(tag));
                if (invalidTags.length > 0) {
                    errors.push({
                        type: 'invalidTags',
                        message: `The following tags are not allowed: ${invalidTags.join(', ')}. Allowed tags: ${allowedTags.tags.join(', ')}`,
                        context: { invalidTags, allowedTags: allowedTags.tags },
                    });
                }
            }
        }
        return errors;
    }
    // ============================================================================
    // Tag Description Management
    // ============================================================================
    /**
     * Get all tag descriptions for a repository
     */
    getTagDescriptions() {
        const tagsDir = this.getTagsDir();
        const descriptions = {};
        if (!this.fs.exists(tagsDir)) {
            return descriptions;
        }
        try {
            const files = this.fs.readDir(tagsDir);
            for (const file of files) {
                if (file.endsWith('.md')) {
                    const tagName = file.slice(0, -3); // Remove .md extension
                    const filePath = this.fs.join(tagsDir, file);
                    try {
                        const content = this.fs.readFile(filePath);
                        descriptions[tagName] = content.trim();
                    }
                    catch (error) {
                        console.error(`Error reading tag description for ${tagName}:`, error);
                    }
                }
            }
        }
        catch (error) {
            console.error('Error reading tags directory:', error);
        }
        return descriptions;
    }
    /**
     * Save a tag description
     */
    saveTagDescription(tag, description) {
        const config = this.getConfiguration();
        // Check description length against tagDescriptionMaxLength
        if (description.length > config.limits.tagDescriptionMaxLength) {
            throw new Error(`Tag description exceeds maximum length of ${config.limits.tagDescriptionMaxLength} characters. ` +
                `Current length: ${description.length}`);
        }
        // Ensure tags directory exists
        const tagsDir = this.getTagsDir();
        this.fs.createDir(tagsDir);
        // Write the description to a markdown file
        const tagFile = this.fs.join(tagsDir, `${tag}.md`);
        this.fs.writeFile(tagFile, description);
    }
    /**
     * Delete a tag description
     */
    deleteTagDescription(tag) {
        const tagsDir = this.getTagsDir();
        const tagFile = this.fs.join(tagsDir, `${tag}.md`);
        if (this.fs.exists(tagFile)) {
            try {
                this.fs.deleteFile(tagFile);
                return true;
            }
            catch (error) {
                console.warn(`Failed to delete tag description for ${tag}:`, error);
                return false;
            }
        }
        return false;
    }
    /**
     * Remove a tag from all notes in the repository
     */
    removeTagFromNotes(repositoryRootPath, tag) {
        const notesWithPaths = this.readAllNotes(repositoryRootPath);
        let modifiedCount = 0;
        for (const noteWithPath of notesWithPaths) {
            const note = noteWithPath.note;
            if (note.tags.includes(tag)) {
                // Remove the tag from the note
                note.tags = note.tags.filter((t) => t !== tag);
                // Save the updated note - reconstruct the note path from the ID and timestamp
                const notePath = this.getNotePath(repositoryRootPath, note.id, note.timestamp);
                this.fs.writeFile(notePath, JSON.stringify(note, null, 2));
                modifiedCount++;
            }
        }
        return modifiedCount;
    }
    /**
     * Replace a tag with another tag in all notes in the repository
     */
    replaceTagInNotes(repositoryRootPath, oldTag, newTag) {
        const notesWithPaths = this.readAllNotes(repositoryRootPath);
        let modifiedCount = 0;
        for (const noteWithPath of notesWithPaths) {
            const note = noteWithPath.note;
            if (note.tags.includes(oldTag)) {
                // Replace the old tag with the new tag
                note.tags = note.tags.map((t) => (t === oldTag ? newTag : t));
                // Remove duplicates if the new tag was already present
                note.tags = [...new Set(note.tags)];
                // Save the updated note - reconstruct the note path from the ID and timestamp
                const notePath = this.getNotePath(repositoryRootPath, note.id, note.timestamp);
                this.fs.writeFile(notePath, JSON.stringify(note, null, 2));
                modifiedCount++;
            }
        }
        return modifiedCount;
    }
    /**
     * Get all used tags for a path
     */
    getUsedTagsForPath(targetPath) {
        const rootPath = '';
        const notes = this.getNotesForPath(targetPath, rootPath, true);
        const counts = new Map();
        for (const n of notes) {
            for (const tag of n.note.tags) {
                counts.set(tag, (counts.get(tag) || 0) + 1);
            }
        }
        return [...counts.entries()].sort((a, b) => b[1] - a[1]).map(([t]) => t);
    }
    /**
     * Get all tags with their descriptions (tags that have description files)
     */
    getTagsWithDescriptions() {
        const descriptions = this.getTagDescriptions();
        const tags = [];
        // Return all tags that have descriptions (these are the available/allowed tags)
        for (const [name, description] of Object.entries(descriptions)) {
            tags.push({ name, description });
        }
        return tags;
    }
    /**
     * Get repository guidance
     */
    getRepositoryGuidance(repositoryPath) {
        try {
            const guidanceFile = this.fs.join(repositoryPath, ALEXANDRIA_DIRS.PRIMARY, 'note-guidance.md');
            // Try to read repository-specific guidance first
            if (this.fs.exists(guidanceFile)) {
                return this.fs.readFile(guidanceFile);
            }
            // Note: Default template would need to be handled differently in pure-core
            // For now, return null if no custom guidance exists
            return null;
        }
        catch {
            return null;
        }
    }
    /**
     * Get allowed tags configuration
     */
    getAllowedTags() {
        const config = this.getConfiguration();
        const enforced = config.tags?.enforceAllowedTags || false;
        if (enforced) {
            // Auto-populate from tags directory
            const tagDescriptions = this.getTagDescriptions();
            const tags = Object.keys(tagDescriptions);
            return { enforced, tags };
        }
        return { enforced, tags: [] };
    }
    // ============================================================================
    // Note Review System
    // ============================================================================
    /**
     * Get all unreviewed notes for a repository path
     */
    getUnreviewedNotes(repositoryPath) {
        const rootPath = '';
        const notes = this.getNotesForPath(repositoryPath, rootPath, true);
        return notes.map((n) => n.note).filter((note) => !note.reviewed);
    }
    /**
     * Mark a note as reviewed by its ID
     */
    markNoteReviewed(repositoryPath, noteId) {
        const note = this.getNoteById(repositoryPath, noteId);
        if (!note) {
            return false;
        }
        // Update the reviewed flag
        note.reviewed = true;
        // Write the updated note back to its file
        const notePath = this.getNotePath(repositoryPath, note.id, note.timestamp);
        this.fs.writeFile(notePath, JSON.stringify(note, null, 2));
        return true;
    }
    /**
     * Mark all notes as reviewed for a given path
     */
    markAllNotesReviewed(repositoryPath) {
        const rootPath = '';
        const notes = this.getNotesForPath(repositoryPath, rootPath, true);
        let count = 0;
        for (const noteWithPath of notes) {
            const note = noteWithPath.note;
            if (!note.reviewed) {
                note.reviewed = true;
                const notePath = this.getNotePath(repositoryPath, note.id, note.timestamp);
                this.fs.writeFile(notePath, JSON.stringify(note, null, 2));
                count++;
            }
        }
        return count;
    }
    // ============================================================================
    // Note Operations
    // ============================================================================
    /**
     * Merge multiple notes into a single consolidated note
     */
    mergeNotes(repositoryPath, input) {
        // Create the merged note with metadata about the merge
        const mergedNoteData = {
            note: input.note,
            directoryPath: repositoryPath,
            anchors: [...new Set(input.anchors)], // Deduplicate anchors
            tags: [...new Set(input.tags)], // Deduplicate tags
            codebaseViewId: input.codebaseViewId,
            metadata: {
                ...input.metadata,
                mergedFrom: input.noteIds,
                mergedAt: new Date().toISOString(),
            },
        };
        const savedNote = this.saveNote(mergedNoteData);
        let deletedCount = 0;
        if (input.deleteOriginals !== false) {
            // Default to true
            for (const noteId of input.noteIds) {
                if (this.deleteNoteById(repositoryPath, noteId)) {
                    deletedCount++;
                }
            }
        }
        return {
            mergedNote: savedNote.note,
            deletedCount,
        };
    }
    /**
     * Get notes for a path with count limit
     * Note: Token limiting requires external dependencies and is not supported in pure-core
     */
    getNotesForPathWithLimit(targetPath, includeParentNotes, limitType, limit) {
        // Get all notes first
        const rootPath = '';
        const allNotes = this.getNotesForPath(targetPath, rootPath, includeParentNotes);
        // Apply count-based limiting
        const results = allNotes.slice(0, Math.max(1, limit));
        return { notes: results };
    }
}
