// @a24z/core-library/dist/pure-core/utils/validation.js
/**
 * Pure validation utilities
 * Platform-agnostic validation message handling
 */
import { DEFAULT_VALIDATION_MESSAGES } from '../types/validation.js';
import { ALEXANDRIA_DIRS } from '../../constants/paths.js';
/**
 * Get the validation messages file path
 */
export function getValidationMessagesPath(fs, repositoryPath) {
    return fs.join(repositoryPath, ALEXANDRIA_DIRS.PRIMARY, 'validation-messages.json');
}
/**
 * Load validation messages from repository
 */
export function loadValidationMessages(fs, repositoryPath) {
    try {
        const messagesPath = getValidationMessagesPath(fs, repositoryPath);
        if (!fs.exists(messagesPath)) {
            return null;
        }
        const content = fs.readFile(messagesPath);
        const messages = JSON.parse(content);
        // Validate the structure
        if (typeof messages !== 'object' || messages === null) {
            return null;
        }
        // Convert string templates to functions
        const overrides = {};
        for (const [key, template] of Object.entries(messages)) {
            if (typeof template === 'string') {
                // Create a function that interpolates the template
                overrides[key] = (data) => {
                    let result = template;
                    // Simple template interpolation for ${variable} patterns
                    for (const [dataKey, value] of Object.entries(data)) {
                        const pattern = new RegExp(`\\$\\{${dataKey}\\}`, 'g');
                        result = result.replace(pattern, String(value));
                    }
                    return result;
                };
            }
        }
        return overrides;
    }
    catch {
        return null;
    }
}
/**
 * Save validation messages to repository
 */
export function saveValidationMessages(fs, repositoryPath, messages) {
    const messagesPath = getValidationMessagesPath(fs, repositoryPath);
    // Ensure alexandria directory exists
    const alexandriaDir = fs.join(repositoryPath, ALEXANDRIA_DIRS.PRIMARY);
    if (!fs.exists(alexandriaDir)) {
        fs.createDir(alexandriaDir);
    }
    // Convert function overrides to string templates for storage
    const templates = {};
    for (const [key, func] of Object.entries(messages)) {
        if (typeof func === 'function') {
            // Store a placeholder template - in practice, these would be provided as strings
            templates[key] = `Custom message for ${key}`;
        }
    }
    // Write the messages file
    fs.writeFile(messagesPath, JSON.stringify(templates, null, 2));
}
/**
 * Message formatter that combines default and custom messages
 */
export class ValidationMessageFormatter {
    messages;
    constructor(overrides) {
        this.messages = {
            ...DEFAULT_VALIDATION_MESSAGES,
            ...overrides,
        };
    }
    format(type, data) {
        const formatter = this.messages[type];
        if (!formatter) {
            throw new Error(`Unknown validation message type: ${type}`);
        }
        return formatter(data);
    }
}
