// @a24z/core-library/dist/node-adapters/NodeFileSystemAdapter.js
/**
 * Node.js implementation of FileSystemAdapter
 *
 * This adapter uses Node.js built-in fs and path modules.
 * It should only be imported in Node.js environments.
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
/**
 * Node.js implementation using built-in fs and path modules
 */
export class NodeFileSystemAdapter {
    constructor() { }
    exists(path) {
        return fs.existsSync(path);
    }
    readFile(path) {
        return fs.readFileSync(path, 'utf8');
    }
    writeFile(path, content) {
        // Ensure parent directory exists
        const dir = this.dirname(path);
        if (!this.exists(dir)) {
            this.createDir(dir);
        }
        // Write to temp file first, then rename (atomic write)
        const tmp = `${path}.tmp`;
        fs.writeFileSync(tmp, content, { encoding: 'utf8' });
        fs.renameSync(tmp, path);
    }
    deleteFile(path) {
        if (this.exists(path)) {
            fs.unlinkSync(path);
        }
    }
    createDir(path) {
        if (!this.exists(path)) {
            fs.mkdirSync(path, { recursive: true });
        }
    }
    readDir(path) {
        if (!this.exists(path))
            return [];
        try {
            const entries = fs.readdirSync(path, { withFileTypes: true });
            return entries.map((entry) => entry.name);
        }
        catch {
            return [];
        }
    }
    isDirectory(path) {
        if (!this.exists(path))
            return false;
        try {
            return fs.statSync(path).isDirectory();
        }
        catch {
            return false;
        }
    }
    deleteDir(path) {
        if (this.exists(path)) {
            // Only delete if empty
            const files = fs.readdirSync(path);
            if (files.length === 0) {
                fs.rmdirSync(path);
            }
        }
    }
    join(...paths) {
        return path.join(...paths);
    }
    dirname(pathString) {
        return path.dirname(pathString);
    }
    relative(from, to) {
        return path.relative(from, to);
    }
    isAbsolute(pathString) {
        return path.isAbsolute(pathString);
    }
    // Repository operations
    normalizeRepositoryPath(inputPath) {
        return normalizeRepositoryPath(inputPath);
    }
    findProjectRoot(inputPath) {
        return this.normalizeRepositoryPath(inputPath);
    }
    getRepositoryName(repositoryPath) {
        const segments = repositoryPath.split(path.sep).filter((s) => s);
        return segments[segments.length - 1] || 'root';
    }
}
export function findGitRoot(startPath) {
    let current = path.resolve(startPath);
    // Handle file paths by starting from the directory
    if (fs.existsSync(current) && fs.statSync(current).isFile()) {
        current = path.dirname(current);
    }
    while (current !== path.dirname(current)) {
        if (fs.existsSync(path.join(current, '.git'))) {
            return current;
        }
        current = path.dirname(current);
    }
    return null;
}
/**
 * Normalize a repository path by finding the git root or project root
 */
export function normalizeRepositoryPath(inputPath) {
    // 1. Try to find git root from the input path
    const gitRoot = findGitRoot(inputPath);
    if (!gitRoot) {
        throw Error('Unable to find Git Root For Path' + inputPath);
    }
    return gitRoot;
}
