// @a24z/core-library/dist/index.js
/**
 * @a24z/core-library
 *
 * Core library exports for the a24z ecosystem.
 * Provides essential functionality for managing notes, views, and configurations.
 */
export { extractCodebaseViewSummary, extractCodebaseViewSummaries, } from './pure-core/types/summary.js';
export { NodeFileSystemAdapter } from './node-adapters/NodeFileSystemAdapter.js';
export { InMemoryFileSystemAdapter } from './node-adapters/InMemoryFileSystemAdapter.js';
// Primary API classes
export { MemoryPalace } from './MemoryPalace.js';
// Project management
export { ProjectRegistryStore } from './projects-core/ProjectRegistryStore.js';
export { AlexandriaOutpostManager } from './projects-core/AlexandriaOutpostManager.js';
// Store exports for direct access if needed
export { CodebaseViewsStore, generateViewIdFromName } from './pure-core/stores/CodebaseViewsStore.js';
// Utilities and rules
export { LibraryRulesEngine } from './rules/index.js';
export { OverviewPathAutoFix } from './pure-core/autofixes/OverviewPathAutoFix.js';
export { ConfigValidator } from './config/validator.js';
// Constants
export { ALEXANDRIA_DIRS } from './constants/paths.js';
export { CONFIG_FILENAME } from './config/schema.js';
// Project utilities
export { getGitRemoteUrl } from './projects-core/utils.js';
export { hasAlexandriaWorkflow, hasMemoryNotes } from './projects-core/workflow-utils.js';
